### PR TITLE
7620: Start removing dependencies on requests

### LINKED ISSUE #7620: Make Requests library an optional dependency
The Requests library seems to be legally problematic so it was suggested to get rid of it as a dependency.

Quoting @mistercrunch :

> It appears it's used
in its simplest form in a few places (requests.get) which should be easy to
replace with urrlib2. The Druid connector uses a more advanced feature
(basic auth), but that could be made just an optional dependency, as we're
deprecating the Druid connector, to be replaced by the SQLAlchemy Druid
connector/dialect. There it's just a matter of moving `pydruid` and
`requests` to the "extra_requires" section of our setup.py, and doing late
imports.

CC: @villebro

### PR DESCRIPTION
**Pull request by sturmer**
### CATEGORY
- [ ] Bug Fix
- [ ] Enhancement (new features, refinement)
- [x] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Make dependency on requests and pydruid connector optional. Note: the first commit is @villebro 's work.

### TEST PLAN
It might require some `@unittest.skipUnless`, I will check the result of the Travis tests first.

### ADDITIONAL INFORMATION
Fixes #7620.

- [x] Has associated issue: 7620.
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@mistercrunch , @villebro

### CODE REVIEW

**Comment by sturmer** [comment]
@mistercrunch @villebro Do you want to have a look? Unless I missed something this should be it.

**Comment by villebro** [comment]
awesome @sturmer ! I will review and test today.

**Comment by villebro** [comment]
Two problems:
1) When I tried to load examples, I got an exception on `superset/connectors/druid/models.py`, which has an import on `requests`. This needs to be wrapped in a try/catch.
2) flask-appbuilder has a dependency on `prison`, which in turn has a dependency on `requests`. I believe this dependency was introduced recently.

The first observation is easy to fix, but the second is more difficult.. @dpgaspar @betodealmeida do you think requests can be replaced with `urllib` on FAB/Prison?

**Comment by sturmer** [comment]
Thanks Ville, I'll check that out today.

**Comment by sturmer** [comment]
I've had a look at Beto's fork of prison and I can't see any usage of requests 🤔 
I've opened a [PR proposing its removal], waiting for comments!

(In the meanwhile, I'll try to fix the dependencies in setup.py and see where I get.)

[PR proposing its removal]: https://github.com/betodealmeida/python-rison/pull/2

**Comment by sturmer** [comment]
Note: This last commit won't work until prison stops depending on requests.

File: superset/connectors/druid/models.py
Comment (villebro) [inline superset/connectors/druid/models.py]: A few observations:
- This code is run at least when `superset load examples` is executed, even when not connected to a druid cluster, causing an `exit(1)` every time. I think this needs to be refactored so that this code is only called when needed, e.g. in the specific code blocks where they are needed.
- The error message should probably be wrapped in a `_('message')` to make it translatable. Furthermore, if the message is wrapped in triple quotes, the leading spaces on the next rows will be included in the message. I believe linters will prefer single quotes over many rows. So something like this:
```python
try:
    ...
except Import Error:
    print(_('pydruid and requests are optional dependencies, and '
            'it seems they are not installed on your system. Aborting...'))
))
```

File: superset/utils/core.py
Comment (villebro) [inline superset/utils/core.py]: Same as above.

**Review by villebro — CHANGES_REQUESTED** [review]

**Comment by betodealmeida** [comment]
@villebro 

> flask-appbuilder has a dependency on prison, which in turn has a dependency on requests. I believe this dependency was introduced recently.

I just published `prison==0.1.2`, without the dependency, thanks to @sturmer! :)

**Comment by sturmer** [comment]
@villebro I've started messing around with the try-caught imports and it seems the business gets quite deep. For instance, if I don't correctly `from pydruid.utils.having import Having`, then I can't declare DimSelector. If I don't declare DimSelector, then I can't use `_get_having_obj(self, col, op, eq)` in models.py. And of course it starts getting deeper than that -- I haven't followed it deep enough but I suspect the whole Python file becomes useless.

Would it make sense to remove completely the druid-related code from the `db upgrade` and `load_examples` commands (whatever that means)? Or you just suggest to `pass` every time we can't correctly declare a symbol for lack of imports?

**Comment by villebro** [comment]
@sturmer I agree, it's not trivial and there are several options, all with pros and cons. Let me digest this for a while and get back with a proposal.

**Comment by villebro** [comment]
Btw, something that can be done now is a PR for https://github.com/dpgaspar/Flask-AppBuilder , i.e. bumping `prison` and removing `requests` there also. FYI @dpgaspar

**Comment by mistercrunch** [comment]
Here: https://github.com/dpgaspar/Flask-AppBuilder/pull/1040

**Comment by sturmer** [comment]
> I haven't followed it deep enough but I suspect the whole Python file becomes useless.
Well in the end it didn't go so deep after all. I've managed to have `superset load_examples` and `superset db upgrade` end successfully (at least without emitting errors), now waiting for the results from Travis.

**Comment by sturmer** [comment]
Hm, if I understand correctly this build won't work until the bump of prison is merged into fab (and the new resulting fab version is specified as a dependency).
image

File: superset/utils/core.py
Comment (sturmer) [reply superset/utils/core.py]: Done.

File: superset/connectors/druid/models.py
Comment (sturmer) [reply superset/connectors/druid/models.py]: Done.

**Comment by villebro** [comment]
@sturmer fab 2.1.5 is now in master

**Comment by sturmer** [comment]
Just restarted the build. Fingers crossed :)

**Comment by sturmer** [comment]
It's time to heed the advice by Maxime and use the ideas from [here](https://github.com/apache/incubator-superset/pull/7605/files#diff-a3e90139d93a6d9706419011538cdcd2R178) to skip tests that depend on PyDruid & co. Will work on that tomorrow.

**Comment by villebro** [comment]
@sturmer I have a few ideas of how to do the conditional importing in a simpler way. Should I make a PR to your branch? Optionally if I can have write access to the branch I could contribute commits, making it easier to collaborate.

**Comment by mistercrunch** [comment]
Sorry about the conflicts, we just merged a large PR running `black` on the whole codebase. It should be easy to apply `black` on your branch and rebase. Let us know if it's not straightforward.

**Comment by sturmer** [comment]
@villebro Sure, I thought you could push to this branch as a project maintainer. If you aren't, do you happen to know how I can authorize you? (I've googled it briefly and couldn't find anything relevant.)

@mistercrunch No worries, I'll solve the conflicts.

**Comment by sturmer** [comment]
@villebro All checks have passed! :)

**Comment by villebro** [comment]
Great @sturmer ! I will review this in detail tonight. I've never made changes to someone else's PR before, so unfamiliar with the process. If there is a need for me to make commits I will look into how it's done.

File: superset/connectors/druid/models.py
Comment (villebro) [inline superset/connectors/druid/models.py]: On second thought I would probably just `pass` here; since `pydruid` is optional (along with `requests`), I don't think it's necessary to state that they are missing.

File: superset/connectors/druid/models.py
Comment (villebro) [inline superset/connectors/druid/models.py]: Same here.

Comment (villebro) [inline superset/connectors/druid/models.py]: Same here.

File: superset/utils/core.py
Comment (villebro) [inline superset/utils/core.py]: For clarity, it might be better to have this try-except block done together with the import, i.e. first trying to import `Having`, and if that's found, continue with defining `DimSelector`. Right now they are quite far apart, and makes understanding the optionality slightly less obvious.

File: superset/connectors/druid/models.py
Comment (villebro) [inline superset/connectors/druid/models.py]: It would probably be good to use either `ImportError` or `ModuleNotFoundError` everywhere? As `ModuleNotFoundError` is a child of `ImportError` and thus more precise, I would opt for that.

File: superset/connectors/druid/models.py
Comment (villebro) [inline superset/connectors/druid/models.py]: Same here ie. `ImportError` vs `ModuleNotFoundError`.

File: superset/connectors/druid/models.py
Comment (villebro) [inline superset/connectors/druid/models.py]: Same here about grouping together with import.

File: superset/utils/core.py
Comment (villebro) [inline superset/utils/core.py]: Same here.

**Review by villebro — CHANGES_REQUESTED** [review]
First pass with comments, but looking pretty good. I will do some testing on my dev env later. It would also be cool to remove the obvious Druid stuff from the UI if dependencies are not present. I can look into that if needed.

File: superset/connectors/druid/models.py
Comment (sturmer) [reply superset/connectors/druid/models.py]: I agree with being uniform but I'm afraid that using `ModuleNotFoundError` doesn't work everywhere, e.g. when importing `Having` if I try to catch `ModuleNotFoundError` it crashes with `ImportError`. I'd try with `ImportError` everywhere if you agree.

File: superset/utils/core.py
Comment (sturmer) [reply superset/utils/core.py]: I've hoisted the class declaration and put it first thing after the imports. In theory, it's doable to also declare the class inside the `try` block, but I'm afraid of flake. Or maybe you meant something else?

File: superset/connectors/druid/models.py
Comment (villebro) [reply superset/connectors/druid/models.py]: Ok, then it probably makes sense to keep it as it is.

**Comment by sturmer** [comment]
Note to self: "TODO - remove the obvious Druid stuff from the UI if dependencies are not present".

**Comment by sturmer** [comment]
@villebro I've had a go at removing Druid-related menu items. Can you please advise on which other places in the UI can be affected, if any?

**Comment by villebro** [comment]
Thanks, I will review later today. It seems the Druid UI removal work might be more involved than I had originally thought, so perhaps that should be saved for a later PR. We'll see.

**Comment by villebro** [comment]
Hmm, I'm kind of split on this one, but starting to lean against just removing the backend stuff should be enough and deal with the frontend later. Removing the UI stuff is a much bigger can of worms than I had anticipated, and there is major risk of feature creep that will derail the originally neatly limited scope of just getting a few deps out of `setup.py` and `requirements.txt`. I vote for removing the last commit and going back to the last one that cleared CI.

One thing that might clean up the code somewhat is to try to import `pydruid` at the earliest possible moment, e.g. `superset/__init__.py` and setting an app config flag as false if `pydruid` isn't present. Later on one would only need to check that, instead of doing a bunch of trying and catching. Not sure about this one, and would now just try to finish this PR as quickly as possible, but if you feel this might have merit and if it's quick to implement then it might make the code slightly cleaner.

**Comment by sturmer** [comment]
Sorry for the long hiatus (vacation), will get back to work today.

**Comment by sturmer** [comment]
I've tried to play a bit with config but it seems to break in a few places and I can't see a clear solution yet. I want to see if Travise passes this commit. Later we can discuss how worth it is to spend more time on using the config-based approach.

**Comment by villebro** [comment]
Thanks @sturmer, let me know if I can be of help!

**Comment by sturmer** [comment]
@villebro as it happens :) I've created a draft PR with my attempt at using configuration [here](https://github.com/apache/incubator-superset/pull/7873). If you have some spare time and want to have a look, you might let me know if you think it can get somewhere or I should just drop it.

**Comment by villebro** [comment]
Thanks will take a look 👍

**Comment by sturmer** [comment]
I'm running black locally but I don't get the errors I get from Travis. Investigating...
(To run it, I enter my virtualenv where python has version 3.6.3, do `export TOXENV=black`, upgrade pip, then run `tox` -- but no error is found.)

**Comment by arpit-agarwal** [comment]
[This] (https://help.github.com/en/articles/inviting-collaborators-to-a-personal-repository) may help on adding collaborators.

**Comment by sturmer** [comment]
@villebro Apart from black failing in obscure (pun intended) ways (I don't manage to reproduce the failure locally), this should be what we're looking for. What do you think?

**Comment by villebro** [comment]
Thanks @sturmer ; I've been slightly busier than I had expected but will review/help out in the coming days!

**Comment by villebro** [comment]
There seems to be something off with your black installation, I got the same changes as Travis:
[diff.txt](https://github.com/apache/incubator-superset/files/3413628/diff.txt). Just do `git apply diff.txt` and you should be good to go. I tested this locally, and seems everything works just fine, and I'm inclined to think this is pretty complete 👍 Can you apply the diff so we can see if CI gives a clean bill of health?

**Comment by sturmer** [comment]
@villebro 
Here's the thing: I've applied the patch, but then flake started complaining. So I removed a few newlines between imports. At that point, black started complaining again...

I'll try to reintroduce the newlines and add #noqa so that flake will shut up. I accept suggestions :)

**Comment by sturmer** [comment]
It works!

**Comment by villebro** [comment]
Great! Really strange that you had to disable `I202`, as it hasn't been done anywhere else in the codebase. Tried to run your branch through local testing with `tox`, but seems my devbox is slightly messed up again, so wasn't able to do that. Let me do some further testing and re-review once more, and if all is well let's put it forward for review to a wider audience.

**Comment by sturmer** [comment]
Dear @villebro, any chance you could have a look? :)

**Comment by villebro** [comment]
Sorry @sturmer, life got in the way, I'll complete my review and testing today!

**Comment by sturmer** [comment]
No worries! ;)

**Comment by villebro** [comment]
Some general comments:
- I clicked around the application as much as I could, including reloading the examples data, but was unable to produce any messy errors. So functionally this seems solid.
- The `try`/`except` additions in `connectors/druid/models.py` still feel slightly messy, but I feel they should be acceptable given the tradeoff of being able to drop problematic licenses. This is also something I feel we can get improve later if needed. Here I think @mistercrunch @john-bodley @betodealmeida could weigh in with comments on whether or not the current solution is appropriate.
- One bigger problem that still needs to be addressed is that the tests that require `pydruid` are now not run as part of CI. Ideally we would install `pydruid` and `requests` when running CI to make sure the Druid tests are run, but I'm unsure what Apache Legal thinks of this. @mistercrunch should we check with them if it's ok to use these deps in our CI setup? I can do the asking if necessary.

**Comment by villebro** [comment]
@sturmer I think it would make sense to rebase this PR; there's been quite a bit of changes to the CI config files, and it would be good to make sure the tests are successful with the new confs. Also, I would suggest adding `pydruid` and `requests` from `requirements.txt` to `requirements-dev.txt`, as CI won't be able to run the tests for Druid otherwise. I'm sure it should be ok to keep the deps in the dev requirements, as installing those is optional (not pulled in by default), and hence should be ok with Apache policies.

**Comment by sturmer** [comment]
Done the rebase. 

Not sure I understand the remark about the `requirements*.txt` @villebro: You mean adding `pydruid` and `requests` to `requirements-dev.txt`? I ask because the way you put it, it seems that you mean _moving_ them from `requirements.txt` to `requirements-dev.txt` while in fact they are absent from the former.

**Comment by villebro** [comment]
Sorry for being unclear. I meant moving as compared to current state of `master`. So the only change necessary would be adding the following lines to `requirements-dev.txt`:
```
pydruid==0.5.6
requests==2.22.0
```

**Comment by sturmer** [comment]
Gotcha & Done!

**Comment by sturmer** [comment]
Weird, it works in my repository https://travis-ci.org/sturmer/incubator-superset
(variation of "but it works on my laptop!" 😄)

**Comment by sturmer** [comment]
I was also thinking: Maybe we could remove the entries in `requirements.txt` which are there `// via requests`, right?

Log: solved merge conflicts, maybe this CI run will be green 🤞

**Comment by sturmer** [comment]
@villebro @mistercrunch @betodealmeida @john-bodley CI green! It is now or never! (Just kidding, take your time -- but if anyone feels like reviewing I'd appreciate feedback. Quoting Ville:
> The try/except additions in connectors/druid/models.py still feel slightly messy, but I feel they should be acceptable given the tradeoff of being able to drop problematic licenses. This is also something I feel we can get improve later if needed. Here I think @mistercrunch @john-bodley @betodealmeida could weigh in with comments on whether or not the current solution is appropriate.

)

**Comment by mistercrunch** [comment]
Nice! Merging while the button is green, we really need to make some progress towards releases!

Started this issue to track blockers:
https://github.com/apache/incubator-superset/issues/7974

**Review by villebro — COMMENTED** [review]
image

**Pull request merged**
